frontend/src/scalars/ui/Chart.vue

```diff
@@ -0,0 +1,598 @@
+            let flattenNumbers = flatten(domainRangeArray);
+            let finalMax = max(flattenNumbers);
+            let finalMin = min(flattenNumbers);
+
+            // Add padding.
+            let PaddedYDomain = this.paddedYDomain(finalMin, finalMax);
+
+            this.setChartOutlierOptions(PaddedYDomain);
+
+            // Store Y domain, if originData is not change, Y domain keep same.
+
+        },
+
+        // Compute  max and min from array, if outlier is true, return quantile range.
+        computeDataRange(arr, isQuantile) {
+            // Get data range.
+            let max;
+            let min;
+            if (!isQuantile) {
+                // Get the orgin data range.
+                max = maxBy(arr, item => item[2])[2];
+                min = minBy(arr, item => item[2])[2];
+            }
+            else {
+                // Get the quantile range.
+                let sorted = sortBy(arr, [item => item[2]]);
+                min = quantile(sorted, minQuantile, item => item[2]);
+                max = quantile(arr, maxQuantile, item => item[2]);
+            }
+            return [min, max];
+        },
+
+        paddedYDomain(min, max) {
+            return {
+                max: max > 0 ? max * 1.1 : max * 0.9,
+                min: min > 0 ? min * 0.9 : min * 1.1
+            };
+        },
+
+        setChartOutlierOptions({min, max}) {
+            this.myChart.setOption({
+                yAxis: {
+                    min,
+                    max
+                }
+            });
+        },
+
+        // Chart horizontal options methods and functions ---- start.
+        setChartHorizon() {
+            let seriesOption = this.myChart.getOption().series;
+            let encodeSeries = val => {
+                return {
+                    encode: {
+                        x: [val]
+                    }
+                };
+            };
+            let stepSeries = seriesOption.map(item => encodeSeries(1));
+            let relativeSeries = seriesOption.map(item => encodeSeries(4));
+            let wallSeries = seriesOption.map(item => encodeSeries(0));
+            let horizontalToxAxisOptions = {
+                step: {
+                    xAxis: {
+                        type: 'value'
+                    },
+                    series: stepSeries
+                },
+                relative: {
+                    xAxis: {
+                        type: 'value'
+                    },
+                    series: relativeSeries
+                },
+                wall: {
+                    xAxis: {
+                        type: 'time'
+                    },
+                    series: wallSeries
+                }
+            };
+            this.myChart.setOption(horizontalToxAxisOptions[this.horizontal]);
+        },
+
+        expandArea() {
+            let pageBoxWidth = document.getElementsByClassName('visual-dl-chart-page-box')[0].offsetWidth;
```

Comment:
You should be using something like `refs` to find elements rendered by this component, instead of DOM query selector lookups.

https://vuejs.org/v2/guide/components.html#Child-Component-Refs

Reply:
very good idea. ```ref``` is the way to go, I will replace all ```document.getElementsBy...``` with ref. However in this case it is finding the parent component element...let me see if there is a way to do it